### PR TITLE
Refactor Registry internals to make concurrent access safe

### PR DESCRIPTION
**Pull request by zcross**
Background: #1109

Related PR(s):
* #1119
* #1124

This adds simple synchronization (`sync.RWMutex`) to accesses of the mutable state within `Registry`. 
It also restructures the data to use a map per entity type, rather than a slice, replacing some linear searches of the data.
As recommended by @sunsingerus , I'm starting here with the Registry and planning on moving onto CHI status (and probably the announcer, at least with a slight change to its builder pattern for chaining log line inputs).

For me, the most pressing open question is whether my read of the code _using_ Registry is semi-accurate: it seems that it is used for internal book keeping, and that individual `ObjectMeta` instances are read in the `Walk...` functions without direct mutation. As I write that, I'm considering whether that was part of the intent of the original code passing these structs by value rather than by pointer. 

Meta: I ran into some issues executing the e2e tests locally (have not fixed yet) and some other simpler (fixed) problems developing this branch on an M1 Mac. I don't want to pull the fixes I made locally into the discussion here, but I won't be shocked if the e2e tests reveal some issue I'm not yet aware of due to this dev environment issue.

## Important items to consider before making a Pull Request
Please check items PR complies to:
* [x] All commits in the PR are squashed. [More info](https://github.com/Altinity/clickhouse-operator/blob/master/CONTRIBUTING.md#how-to-make-pr) 
* [x] The PR is made into dedicated `next-release` branch, **not into** `master` branch<sup>1</sup>. [More info](https://github.com/Altinity/clickhouse-operator/blob/master/CONTRIBUTING.md#how-to-make-pr)
* [x] The PR is signed. [More info](https://github.com/Altinity/clickhouse-operator/blob/master/CONTRIBUTING.md#sign-your-work)

### CODE REVIEW

File: pkg/model/registry.go [hunk 1]
```diff
@@ -293,48 +335,99 @@ func (r *Registry) isEqual(a, b v1.ObjectMeta) bool {
...
+	// and then begin operating on that (it uses a finer grained lock).
+	setForType := r.getOrCreateObjSet(entityType)
+
+	// Having acquired the type-specific ObjectMeta set, return the result of removal success.
+	return setForType.remove(&meta)
+}
+
+// getOrCreateObjSet resolves the singleton objectMetaSet for this registry, of the given entityType.
+// This uses a coarse grained lock on the entire registry.
+func (r *Registry) getOrCreateObjSet(entityType EntityType) *objectMetaSet {
```
Comment (sunsingerus) [inline pkg/model/registry.go:348]: We have an internal agreement that functions with behavior like 'get or create if not exists and then get' are called either `Ensure()` or starting with `Ensure`, like `Ensure<Something>()`
Ex.:
```
func (chi *ClickHouseInstallation) EnsureStatus() *ChiStatus {
```
or
```
func (o *StatefulSetPollOptions) Ensure() *StatefulSetPollOptions {
```
Most likely this is the case

**Review by sunsingerus — CHANGES_REQUESTED** [review]

**Comment by zcross** [comment]
> > ... Registry ... is used for internal book keeping, and that individual ObjectMeta instances are read ... without direct mutation
> 
> Yes, this is correct.
> 
> Main idea of the `Registry` is to keep registry of reconciled objects for two main reasons (at least for now)
> 
>     1. To clear (delete) all objects that belong to the CHI but not listed in registry as reconciled - thus removing all remnants of previous iterations of the `CHI` w/o knowing what they are. Clean all what is not listed in the registry as belonging to current iteration of the `CHI`
> 
>     2. Statistics and counting
> 
> 
> Entities inside `Registry` are not mutated, but are read only, at least for now

Perfect, I think that's compatible with the approach taken in this PR in that case.

File: pkg/model/registry.go (hunk 1 above)
Comment (zcross) [reply pkg/model/registry.go]: (Addressed)

**Comment by zcross** [comment]
Meta: I understand that you want commits squashed. I tend to incrementally commit to address PR feedback, so that the feedback-inspired changes are visible at the commit level while the PR is under review, and then squashing to one commit before merging. Does that work, or do you prefer squashing aggressively during the review process? Thanks!

**Comment by sunsingerus** [comment]
> do you prefer squashing aggressively during the review process?

No need to squash immediately, this can be done once at the very end

**Comment by sunsingerus** [comment]
> ... project seems to rely on end to end behavioral tests. ...

The intention is to cover the whole functionality (or at least as much as possible) with e2e tests. Tests are typically added along with new functionality. However, in case any strange / corner case behavior is found later, tests are modified to cover it. Generally speaking, successful run of the test suite gives good hope that the operator can do most of the typical use-case scenarios.

> ... I was wondering about dropping in some very simple go tests ...

Yes, sure. Tests covering any functionality (new and already existing, and especially complex scenarios) are always welcome.

File: pkg/model/registry_test.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+var testPvcD = v1.ObjectMeta{
+	Name:      "persistentvolumeclaim-D",
+	Namespace: testNamespace,
+}
+
+// NB: These tests mostly exist to exercise synchronization and detect regressions related to them via the
+// Golang race detector. See: https://go.dev/blog/race-detector
+// In short, add -race to the go test flags when running this.
+func Test_Registry_BasicOperations_ConcurrencyTest(t *testing.T) {
```
Comment (zcross) [inline pkg/model/registry_test.go:54]: @sunsingerus here's the minimal (not assertion-heavy) go test I mentioned, its purpose being to exercise the synchronization primitives and detect any regressions related to concurrency

File: pkg/model/registry.go [hunk 3]
```diff
@@ -293,48 +335,99 @@ func (r *Registry) isEqual(a, b v1.ObjectMeta) bool {
...
+	r.RUnlock()
+	if ok {
+		return existing
 	}
 
-	return false
+	// create case (requires write lock)
+	newSet := newObjectMetaSet(entityType)
+	r.Lock()
+	r.r[entityType] = newSet
```
Comment (zcross) [inline pkg/model/registry.go:364]: this should actually check for existence again, will fix

**Comment by zcross** [comment]
Squashing this as I await feedback and repointing the other open PR to this as its base so I can branch a third stacked PR for the actual reconciliation logic changes.
